pkg/logs/logs.go

```diff
@@ -30,37 +37,47 @@ func Start() error {
 	return nil
 }
 
-// run sets up the pipeline to process logs and them to Datadog back-end
+// run sets up the pipeline to process logs and send them to Datadog back-end
 func run() {
 	isRunning = true
 
-	cm := sender.NewConnectionManager(
+	connectionManager := sender.NewConnectionManager(
 		config.LogsAgent.GetString("log_dd_url"),
 		config.LogsAgent.GetInt("log_dd_port"),
 		config.LogsAgent.GetBool("dev_mode_no_ssl"),
 	)
 
-	auditorChan := make(chan message.Message, config.ChanSizes)
-	a := auditor.New(auditorChan)
-	a.Start()
-
-	pp := pipeline.NewProvider()
-	pp.Start(cm, auditorChan)
+	messageChan := make(chan message.Message, config.ChanSize)
+	auditor := auditor.New(messageChan)
+	pipelineProvider := pipeline.NewProvider(connectionManager, messageChan)
 
 	sources := config.GetLogsSources()
 
-	l := listener.New(sources.GetValidSources(), pp)
-	l.Start()
-
-	tailingLimit := config.LogsAgent.GetInt("log_open_files_limit")
-	s := tailer.New(sources.GetValidSources(), tailingLimit, pp, a)
-	s.Start()
-
-	c := container.New(sources.GetValidSources(), pp, a)
-	c.Start()
+	networkListeners := listener.New(sources.GetValidSources(), pipelineProvider)
+	containersScanner := container.New(sources.GetValidSources(), pipelineProvider, auditor)
+	filesScanner := tailer.New(
+		sources.GetValidSources(),
+		config.LogsAgent.GetInt("log_open_files_limit"),
+		pipelineProvider,
+		auditor,
+		tailer.DefaultSleepDuration,
+	)
 
+	restart.Start(auditor, pipelineProvider, networkListeners, containersScanner, filesScanner)
 	status.Initialize(sources.GetSources())
 
+	inputs := restart.NewParallelGroup(filesScanner, containersScanner, networkListeners)
+	agentPipeline = restart.NewSerialGroup(inputs, pipelineProvider, auditor)
+}
+
+// Stop stops properly the logs-agent to prevent data loss
+// All Stop methods are blocking which means that Stop only returns
+// when the whole pipeline is flushed
+func Stop() {
+	if isRunning {
+		log.Info("Stopping logs-agent")
+		agentPipeline.Stop()
```

Comment:
I think there's a race condition at this point (although it's hard to run into it): when `Stop` is called, even if `isRunning` is true, you can't be sure that agentPipeline isn't `nil` since you don't know if `run` has finished. So you might be trying to dereference a `nil` pointer here.

Solution would be either to try stopping `agentPipeline` only if it's not nil (not a great solution, but we do similar things in the agent code), or have a sync mechanism to make sure `run` is finished before attempting to call `agentPipeline.Stop` (better IMO), or make `Start` blocking. WDYT?

Reply:
Hey @olivielpeau , good catch, thanks.
What about moving `isRunning` to the very end of the method `run` ?

Reply:
Works for me! The only caveat is that, in the conditions of this race condition, calling `Stop` on the logs agent would not actually stop it. Still better than the nil pointer dereference though :)